### PR TITLE
feat(sql): added support for unit microseconds (U) in sample by statements #1984.

### PR DESCRIPTION
**Pull request by KevinChen16**
Documentation PR https://github.com/questdb/questdb.io/pull/841

Adding support of microseconds in timestamp sampler(see #1984 ):
- `createTimestampSampler(long interval, char timeUnit, int position)` in TimestampSamplerFactory.java

Adding test cases for the microsecond in timestamp:
- `testMicros()` in TimestampSamplerFactoryTest.java
- `testSampleByMicrosFillNoneNotKeyedEmpty()` in SampleByTest.java

Please let me know whether I write proper test cases. Thanks!

updates:
- modifying `testMicros()` and `testSampleByMicrosFillNoneNotKeyedEmpty()`
- adding `testMillis()` in TimestampSamplerFactoryTest.java

### CODE REVIEW

File: core/src/test/java/io/questdb/griffin/engine/groupby/SampleByTest.java
Comment (marregui) [inline core/src/test/java/io/questdb/griffin/engine/groupby/SampleByTest.java:2565]: This statement does create the table, however the `long_sequence(0)` means that no data are inserted, did you do this in purpose?, what is the reason?, did you mean to simply create the table:

```

create table x (
    a double,
    b symbol,
    k timestamp
) timestamp(k) partition by none
```

and then populate it with the subsequent insert from select statement?.

File: core/src/test/java/io/questdb/griffin/engine/groupby/TimestampSamplerFactoryTest.java
Comment (marregui) [inline core/src/test/java/io/questdb/griffin/engine/groupby/TimestampSamplerFactoryTest.java:112]: micros `k < 1000001`, and then the loop would take forever. May I suggest to use something like:

```
    Rnd rand = new Rnd();
    for (int round = 0; round < 1000; round++) {
         int k = rand.nextInt(1000001);
         ...
``` 

(rounds < sqrt(micros))

**Review by marregui — CHANGES_REQUESTED** [review]

File: core/src/test/java/io/questdb/griffin/engine/groupby/TimestampSamplerFactoryTest.java
Comment (KevinChen16) [reply core/src/test/java/io/questdb/griffin/engine/groupby/TimestampSamplerFactoryTest.java:112]: Thanks! And I think it is possible to add a `testMillis()` method, and both `k < 1001`, right?

File: core/src/test/java/io/questdb/griffin/engine/groupby/SampleByTest.java
Comment (KevinChen16) [reply core/src/test/java/io/questdb/griffin/engine/groupby/SampleByTest.java:2565]: Yeah, you're right! And I'll change my SQL commands.

**Comment by KevinChen16** [comment]
Hi @marregui , I've modified the code, and please check, thanks!

File: core/src/test/java/io/questdb/griffin/engine/groupby/TimestampSamplerFactoryTest.java
Comment (marregui) [inline core/src/test/java/io/questdb/griffin/engine/groupby/TimestampSamplerFactoryTest.java:137]: Thank you @KevinChen16 looking much closer to a merge, although there is now a lot of repeated code which would be nice to refactor so that the tests become more readable. 

Given this new method (please add it):

```
private static TimestampSampler createTimestampSampler(int bucketSize, char units, StringSink sink) throws SqlException {
        sink.clear();
        if (bucketSize > 0) {
            sink.put(bucketSize);
        }
        sink.put(units);
        TimestampSampler sampler = TimestampSamplerFactory.getInstance(sink, 120);
        Assert.assertNotNull(sampler);
        return sampler;
    }
```

would you please refactor TSFT to use it?. Then, for `testMicros()` you are only testing `k` up to 1001, when the micros range is up to 1000001, that is why I suggested trying k values for up to 1000 attempts, where at each attempt k is randomly chosen like this:

```
@Test
    public void testMicros() throws NumericException, SqlException {
        final StringSink sink = new StringSink();
        final long ts = TimestampFormatUtils.parseUTCTimestamp("2022-04-23T10:33:00.123456Z");
        final Rnd rand = new Rnd();
        for (int j = 0; j < 1000; j++) {
            final int k = rand.nextInt(1000001);
            final TimestampSampler sampler = createTimestampSampler(k, 'U', sink);
            final long bucketSize = (k == 0 ? 1 : k);
            final long expectedTs = ts - ts % bucketSize;
            for (int i = 0; i < bucketSize; i+=4) {
                long actualTs = sampler.round(expectedTs + i);
                if (expectedTs != actualTs) {
                    Assert.fail(String.format(
                            "Failed at: %s, i: %d. Expected: %s, actual: %s",
                            sink, i, Timestamps.toString(expectedTs), Timestamps.toString(actualTs))
                    );
                }
            }
        }
    }
```
Could you please amend.

Thank you in advance!!!

**Review by marregui — CHANGES_REQUESTED** [review]

File: core/src/test/java/io/questdb/griffin/engine/groupby/TimestampSamplerFactoryTest.java
Comment (KevinChen16) [reply core/src/test/java/io/questdb/griffin/engine/groupby/TimestampSamplerFactoryTest.java:137]: Sure! I'll upload it soon! Thank you for your advice!

File: core/src/test/java/io/questdb/griffin/engine/groupby/TimestampSamplerFactoryTest.java
Comment (marregui) [inline core/src/test/java/io/questdb/griffin/engine/groupby/TimestampSamplerFactoryTest.java:66]: afaik bucket size should match the value given to the constructor in `TimestampSamplerFactory.createTimestampSampler`, which for this case is:

```
final long bucketSize = Timestamps.MINUTE_MICROS * (k == 0 ? 1 : k);
```

File: core/src/test/java/io/questdb/griffin/engine/groupby/TimestampSamplerFactoryTest.java
Comment (marregui) [inline core/src/test/java/io/questdb/griffin/engine/groupby/TimestampSamplerFactoryTest.java:86]: same logic as stated in previous comment, should be `Timestamps.SECOND_MICROS * k...`

**Review by marregui — CHANGES_REQUESTED** [review]

File: core/src/test/java/io/questdb/griffin/engine/groupby/TimestampSamplerFactoryTest.java
Comment (KevinChen16) [reply core/src/test/java/io/questdb/griffin/engine/groupby/TimestampSamplerFactoryTest.java:66]: So can I try like this? Thanks!
```
final long bucketSize = Timestamps.SECOND_MICROS * (k == 0 ? 1 : k);
final long expectedTs = ts - ts % bucketSize;
for (int i = 0; i < bucketSize; i+=4 * Timestamps.MILLI_MICROS) {
     long actualTs = sampler.round(expectedTs + I);
}
```

File: core/src/test/java/io/questdb/griffin/engine/groupby/TimestampSamplerFactoryTest.java
Comment (KevinChen16) [reply core/src/test/java/io/questdb/griffin/engine/groupby/TimestampSamplerFactoryTest.java:86]: So can I try like this? Thanks!
```
final long bucketSize = Timestamps.SECOND_MICROS * (k == 0 ? 1 : k);
final long expectedTs = ts - ts % bucketSize;
for (int i = 0; i < bucketSize; i+=4 * Timestamps.MILLI_MICROS) {
     long actualTs = sampler.round(expectedTs + I);
}
```

Comment (marregui) [reply core/src/test/java/io/questdb/griffin/engine/groupby/TimestampSamplerFactoryTest.java:86]: I think that if you leave the essence of the test for seconds and micros as it was, except for the use of the `createTimestampSampler` function we will be on the win.

Comment (KevinChen16) [reply core/src/test/java/io/questdb/griffin/engine/groupby/TimestampSamplerFactoryTest.java]: Ok, thanks! I've pulled the newest version, please check it! Appreciate it! @marregui

Comment (KevinChen16) [reply core/src/test/java/io/questdb/griffin/engine/groupby/TimestampSamplerFactoryTest.java]: Hi, I've pulled the newest version, please check, thanks! @marregui

**Comment by marregui** [comment]
hola @KevinChen16 looking good! thank you for your endurance. Could you please update your fork in preparation for merging?, thank you in advance!

**Comment by KevinChen16** [comment]
> hola @KevinChen16 looking good! thank you for your endurance. Could you please update your fork in preparation for merging?, thank you in advance!

OK, thanks!

**Comment by KevinChen16** [comment]
Hi! @marregui I've updated my fork, please let me know if there is any other thing I should do. Thanks!

**Comment by marregui** [comment]
hi @KevinChen16 I have discovered `SqlParser.isValidSampleByPeriodLetter` which suggests some validation, would you be able to check its use in the presence micros? thank you in advance.

**Comment by KevinChen16** [comment]
> hi @KevinChen16 I have discovered `SqlParser.isValidSampleByPeriodLetter` which suggests some validation, would you be able to check its use in the presence micros? thank you in advance.

OK! I think I may find other method like this and modify them together. Thanks!

**Comment by marregui** [comment]
thank you @KevinChen16 !! looking great. In preparation for approval/merge, would you please be able to update the fork and push again. It would be awesome if you could `git rebase -i master` before you push, and left a single commit with a comment like:

```
feat(sql): Add support for unit microseconds (U) in sample by statements #1984.
```

If you feel like it, would you please be able to amend the documentation?. This would be achieved through a new PR on repo https://github.com/questdb/questdb.io to update file https://github.com/questdb/questdb.io/blob/master/docs/reference/sql/sample-by.md  (then would you please edit the description of this PR to add the link to the PR containing the documentation change). 

Thank you in advance.

**Comment by KevinChen16** [comment]
> thank you @KevinChen16 !! looking great. In preparation for approval/merge, would you please be able to update the fork and push again. It would be awesome if you could `git rebase -i master` before you push, and left a single commit with a comment like:
> 
> ```
> feat(sql): Add support for unit microseconds (U) in sample by statements #1984.
> ```
> 
> If you feel like it, would you please be able to amend the documentation?. This would be achieved through a new PR on repo https://github.com/questdb/questdb.io to update file https://github.com/questdb/questdb.io/blob/master/docs/reference/sql/sample-by.md (then would you please edit the description of this PR to add the link to the PR containing the documentation change).
> 
> Thank you in advance.

Sure! Thanks!

**Review by marregui — APPROVED** [review]
LGTM thank you @KevinChen16

**Comment by marregui** [comment]
hey @KevinChen16 I have created a pr https://github.com/questdb/questdb.io/pull/841 for the documentation enhancement, to save you the pain :) . Thank you for this pr!!

**Comment by KevinChen16** [comment]
> hey @KevinChen16 I have created a pr [questdb/questdb.io#841](https://github.com/questdb/questdb.io/pull/841) for the documentation enhancement, to save you the pain :) . Thank you for this pr!!

Thank you!

**Comment by KevinChen16** [comment]
@marregui Hi, I found that it is not being merged now, so I wonder what else should I do, thank you!

**Comment by marregui** [comment]
Hi @KevinChen16 thank you!, we now require two approvals :) I will ask the team to procure the second one.
In the meantime, I am pretty sure they will ask to keep the fork up to date, there were a few merges last night, would you please update?.

**Comment by KevinChen16** [comment]
> Hi @KevinChen16 thank you!, we now require two approvals :) I will ask the team to procure the second one. In the meantime, I am pretty sure they will ask to keep the fork up to date, there were a few merges last night, would you please update?.

OK, thanks!

**Review by bluestreak01 — APPROVED** [review]

**Comment by KevinChen16** [comment]
Thank you @marregui  and @bluestreak01 ! I wonder what should I do after getting two approvals from you to merge the PR? Thank you!

**Comment by bluestreak01** [comment]
Thank you @KevinChen16 ! Not much else to do, we were waiting for CI to do its things. Can merge now.

**Comment by marregui** [comment]
hey @KevinChen16 congrats on merging!
do claim your swag at level two if you haven't already, follow this link https://questdb.io/community/

**Comment by marregui** [comment]
thank you!!!

**Pull request merged**
